### PR TITLE
Implement membership changes

### PR DESCRIPTION
**Pull request by ltseeley**
Adds functionality for leader to listen for changes to the `sawtooth.consensus.raft.peers` setting and trigger a configuration change, and for all nodes to handle/apply configuration changes.

Also adds a Bash script for testing dynamic membership of a Raft network. This script will be adapted to work in a Docker service in a later PR.

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -142,3 +132,17 @@ pub fn peer_id_to_raft_id(peer_id: &PeerId) -> u64 {
...
+        .get("sawtooth.consensus.raft.peers")
+        .expect("'sawtooth.consensus.raft.peers' must be set to use Raft");
+
+    let peers: Vec<String> = serde_json::from_str(peers_setting_value)
+        .expect("Invalid value at 'sawtooth.consensus.raft.peers'");
+
+    peers.into_iter()
+        .map(|s| PeerId::from(hex::decode(s).expect("Peer id not valid hex")))
+        .collect()
+}
```
Comment (aludvik) [inline src/config.rs]: This refactor should be a separate commit.

File: src/node.rs [hunk 2]
```diff
@@ -142,9 +146,48 @@ impl<S: StorageExt> SawtoothRaftNode<S> {
...
+                    info!("Leader({:?}) detected configuration change {:?}", self.peer_id, change);
+                    debug!("Leader({:?}) transition to ChangingConfig", self.peer_id);
+                    self.leader_state = Some(LeaderState::ChangingConfig);
+                    self.raw_node.propose_conf_change(vec![], change).unwrap();
+                });
+            } else {
+                debug!("Leader({:?}) transition to Building block {:?}", self.peer_id, block_id);
+                self.leader_state = Some(LeaderState::Building(Instant::now()));
+                self.service.initialize_block(None).expect("Failed to initialize block");
+            }
```
Comment (aludvik) [inline src/node.rs]: This change should be refactored behind a private method. Something like `check_for_conf_change(&self) -> Option<ConfChange>` would probably be clearer and a good level of abstraction.

File: src/node.rs [hunk 3]
```diff
@@ -289,6 +332,33 @@ impl<S: StorageExt> SawtoothRaftNode<S> {
...
+                        ConfChangeType::AddNode => {
+                            self.raft_id_to_peer_id.insert(
+                                raft_id,
+                                PeerId::from(Vec::from(change.get_context()))
+                            );
+                        },
+                        _ => {}
+                    }
+
+                    self.raw_node.apply_conf_change(&change);
```
Comment (aludvik) [inline src/node.rs]: I would suggest moving L338 to L355 behind a private method, something like `apply_conf_change(change: &ConfChange)` is probably good.

File: tests/test_dynamic_membership.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
```
Comment (aludvik) [inline tests/test_dynamic_membership.sh]: Needs copyright header

File: tests/test_dynamic_membership.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+### This script tests dynamic membership by starting a Raft network with 3 nodes, adding a 4th node, applying a workload, and checking that all 4 nodes reach block 5
+
+echo "Starting initial network"
+docker-compose -f ../adhoc/admin.yaml up -d
+docker-compose -p alpha -f ../adhoc/node.yaml up -d
+docker-compose -p beta -f ../adhoc/node.yaml up -d
+GENESIS=1 docker-compose -p gamma -f ../adhoc/node.yaml up -d
+
+echo "Gathering list of initial keys and REST APIs"
+INIT_KEYS=($(docker exec $(docker ps | grep admin | awk '{print $1;}') bash -c 'cd /shared_data/validators && paste $(ls -1) -d , | sed s/,/\ /g'))
```
Comment (aludvik) [inline tests/test_dynamic_membership.sh]: You can assign `$(docker ps | grep admin | awk '{print $1;}')` to a variable so you don't need to rerun the command constantly, as the output shouldn't change.

File: tests/test_dynamic_membership.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+echo "Initial keys:" ${INIT_KEYS[*]}
+INIT_APIS=($(docker exec $(docker ps | grep admin | awk '{print $1;}') bash -c 'cd /shared_data/rest_apis && ls -d *'))
+echo "Initial APIs:" ${INIT_APIS[*]}
+
+echo "Waiting until network has started"
+docker exec $(docker ps | grep admin | awk '{print $1;}') bash -c 'while true; do if [[ $(sawtooth block list --url "http://$(ls /shared_data/rest_apis | head -n 1):8008" 2>&1) == *"BLOCK_ID"* ]]; then echo "Network ready" && break; else echo "Still waiting..." && sleep 0.5; fi; done;'
```
Comment (aludvik) [inline tests/test_dynamic_membership.sh]: Wrap long lines at 80 characters

File: src/node.rs [hunk 7]
```diff
@@ -289,6 +299,17 @@ impl<S: StorageExt> SawtoothRaftNode<S> {
                 if entry.get_entry_type() == EntryType::EntryNormal {
                     let block_id: BlockId = BlockId::from(Vec::from(entry.get_data()));
                     self.block_queue.add_block_commit(block_id);
+                } else if entry.get_entry_type() == EntryType::EntryConfChange && entry.get_term() != 1 {
+                    let change: ConfChange = protobuf::parse_from_bytes(entry.get_data())
+                        .expect("Failed to parse ConfChange");
```
Comment (peterschwarz) [inline src/node.rs]: Should we crash here? Even though Raft is not byzantine, this could be an interesting attack vector.  Perhaps the change should just be ignored if it can't be parsed, since presumably the change couldn't be parsed by anyone else in the network, either.

File: src/node.rs [hunk 8]
```diff
@@ -333,6 +354,67 @@ impl<S: StorageExt> SawtoothRaftNode<S> {
             }
         }
     }
+
+    fn check_for_conf_change(&mut self, block_id: BlockId) -> Option<ConfChange> {
+        // Get list of peers from settings
+        let settings = self.service
+            .get_settings(block_id, vec![String::from("sawtooth.consensus.raft.peers")])
+            .expect("Failed to get settings");
```
Comment (peterschwarz) [inline src/node.rs]: The error message that was deleted `"'sawtooth.consensus.raft.peers' must be set to use Raft"` seemed more informative.

Comment (ltseeley) [reply src/node.rs]: That error message will be displayed if the peers setting isn't found (in the call to `get_peers_from_settings`). This is a different error where it can't even get the settings; if this fails, it's due to an actual error reported by the service.

File: src/node.rs (hunk 7 above)
Comment (ltseeley) [reply src/node.rs]: I suppose that makes sense to me. @aludvik what are your thoughts on this?

File: src/node.rs (hunk 8 above)
Comment (peterschwarz) [reply src/node.rs]: What happens when the validator goes down?  Does Raft go down too?

Comment (ltseeley) [reply src/node.rs]: Yes, Raft goes down as well.

Comment (peterschwarz) [reply src/node.rs]: It sounds like this is currently a common pattern in sawtooth-raft, when dealing with network communication with the validator.  We should create a story to clean this up and handle any places where an error could occur from the validator gracefully.

File: src/node.rs (hunk 7 above)
Comment (aludvik) [reply src/node.rs]: I think it is better to crash than to carry on in an invalid state. If a setting change occurs, then you won't be able to restart nodes anyways so you are going to eventually end up crashed anyways (they wouldn't be able to determine what world view the rest of the network has). 

On-chain settings are already permissioned so we don't have to worry about a bad actor changing the setting maliciously, just admins making mistakes. I think we can implement mulit-node configuration changes later on as a new feature by extending the behavior of the library at the application level through multiple rounds of configuration changes.

Comment (peterschwarz) [reply src/node.rs]: But that means a unparsable configuration setting (due to a typo) brings down the whole network.  Yes, on chain settings are permissioned. but that doesn't mean a typo couldn't get through the voting process. Plus, every node will receive the same bad setting, so they all should ignore the change.

That's not programmer error, that's user error.  `expect` should be reserved for programmer error, and other truly fatal conditions.

I think this actually applies more to the `get_peers_from_settings` function than this particular line.
